### PR TITLE
New spec for Feat/refactor nonce

### PR DESCRIPTION
**Pull request by huwenqing0606**
This is a new spec we prepared for the feat/refactor_nonce branch. It is not in the final form, but it should have a relatively clear presentation of the major ideas in the design. We hope to use this doc to help developers get the ideas quickly and start to contribute. 
For the first reading, you may skip the <i>italics</i> part annotated as <i>Notes</i> or <i>Discussions</i>, since these are just additional memo. The constraints part are not finished yet and there are other TODO parts in the doc that we will update later. If you are already familiar with the MPT table and the SMTTrace stuff in zktrie, you can also start reading from the section "MPT proofs as witnesses for the feat/refactor branch"

### CODE REVIEW

**Comment by huwenqing0606** [comment]
corresponding HackMD link here https://hackmd.io/@dieGzUCgSGmRZFQ7SDxXCA/H1BGptdfh

File: spec/mpt-proof_feat-refactor.md [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+        1. `None`: the update doesn't touch the account's storage. I.e. the proof type matches `MPTProofType::{NonceChanged, BalanceChanged, CodeHashExists, PoseidonCodeHashExists, CodeSizeExists, AccountDoesNotExist}`
+        2. `[None, None]` the storage slot for the update starts and ends with value 0. Either the proof type is `MPTProofType::StorageDoesNotExist` or it is `MPTProofType::StorageChanged` and the new and old values are both 0. 
+        3. `[Some(...), Some(...)]`: the proof type is `MPTProofType::StorageChanged` and the new and old values are non-zero.
+        4. `[None, Some(...)]`: the proof type is `MPTProofType::StorageChanged` and the old value is 0 and the new value is non-zero.
+        5. `[Some(...), None]`: the proof type is `MPTProofType::StorageChanged` and the old value is non-zero and the new value is non-zero.
+    - `common_state_root`: 32-byte hex, hash. This is optional and is present iff `state_path` and `state_update` are both `None`. (It is not possible for only one of `state_path` or `state_update` to be `None`.)
```
Comment (silathdiir) [inline spec/mpt-proof_feat-refactor.md:82]: May be a typo of `if`?
```suggestion
    - `common_state_root`: 32-byte hex, hash. This is optional and is present if `state_path` and `state_update` are both `None`. (It is not possible for only one of `state_path` or `state_update` to be `None`.)
```

File: spec/mpt-proof_feat-refactor.md [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+(3) <i>insert to/delete from fill</i>: insert an account/storage slot to fill in an empty node (<i>insert to fill</i>) as well as delete an account/storage slot from a filled leafnode and leaves with an empty node (<i>delete from fill</i>). The account/storage under this operation is named <b>type 2</b>, i.e., type 2 account/storage are also empty. Where they could be in the MPT, there is instead an empty node. Figure below (for insert case, and delete case just swap old and new):
+![PathType_Common_insertdelete_fill](https://i.imgur.com/agw0OsP.jpg)
+
+### PathTypes
+
+#### PathType::Common
+
+`PathType::Common` refers to the sitation that the old and new path share the same toplogical configuration.
```
Comment (silathdiir) [inline spec/mpt-proof_feat-refactor.md:206]:
```suggestion
`PathType::Common` refers to the situation that the old and new path share the same toplogical configuration.
```

File: spec/mpt-proof_feat-refactor.md (hunk 1 above)
Comment (lispc) [reply spec/mpt-proof_feat-refactor.md:82]: if and only if?

File: spec/mpt-proof_feat-refactor.md [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+#### PathType::Common
+on rows with `PathType::Common`:
+- for the old common path, `old_hash_previous==poisedon_hash(old_hash_left_child, old_hash_right_child)`
+- for the new common path, `new_hash_previous==poisedon_hash(new_hash_left_child, new_hash_right_child)`
+- for type 2 non-existence proof, common path includes an empty node as leafnode. So we provide constraints for this case
+    - if next row has `PathType::ExtensionNew` and `SegmentType::AccountTrie0`, then current row has `old_hash==0`
+    - if next row has `PathType::ExtensionOld` and `SegmentType::AccountTrie0`, then `new_hash==0`
+
+
+#### PathType::ExtensionNew
```
Comment (silathdiir) [inline spec/mpt-proof_feat-refactor.md:302]: Seems missing `PathType::ExtensionOld` constraints (as `new_value==0` and so on).

Comment (huwenqing0606) [reply spec/mpt-proof_feat-refactor.md:302]: ExtensionOld shall be just a flip of ExtensionNew by swapping old and new. They are symmetric.

**Pull request merged**